File: tools/content_tools/content_tools/validate_docs_links.py
from __future__ import annotations

import argparse
import re
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import unquote

from content_tools import doc_registry, sop_parse


PROCESS_MARKDOWN_GLOBS = [
    "_docs/**/*.md",
    "docs/**/*.md",
    "templates/**/*.md",
]
PROCESS_MARKDOWN_FILES = [
    ".goal-v1.md",
    "PROJECT_PLAN.md",
    "PORTAL_ANALYSIS.md",
    "README.md",
]
MARKDOWN_LINK_RE = re.compile(r"(?P<image>!)?\[[^\]]*]\((?P<target>[^)]+)\)")
WIKI_REF_RE = re.compile(r"\[\[(?P<target>[^\]|]+)(?:\|[^\]]+)?\]\]")
CODE_BLOCK_RE = re.compile(r"```.*?```", re.DOTALL)
INLINE_CODE_RE = re.compile(r"`[^`]+`")
SCHEME_RE = re.compile(r"^[a-z][a-z0-9+.-]*:", re.IGNORECASE)
SOURCE_DOC_IDS_RE = re.compile(r"sourceDocIds\s*:\s*\[(?P<body>.*?)\]", re.DOTALL)
INSTRUCTION_DOC_ID_RE = re.compile(r"instructionDocId\s*:\s*['\"](?P<id>[^'\"]+)['\"]")
STRING_LITERAL_RE = re.compile(r"['\"](?P<value>[^'\"]+)['\"]")
EXTERNAL_DOC_OBJECT_RE = re.compile(
    r"\{\s*"
    r"id:\s*['\"](?P<id>[^'\"]+)['\"]\s*,\s*"
    r"path:\s*['\"](?P<path>[^'\"]+)['\"]\s*,\s*"
    r"reason:\s*['\"](?P<reason>[^'\"]+)['\"]\s*,?\s*"
    r"\}",
    re.DOTALL,
)


@dataclass(frozen=True)
class MarkdownReference:
    source_path: Path
    target: str
    is_image: bool


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate internal DataOps document references.")
    parser.add_argument("--repo-root", default=Path.cwd(), type=Path)
    parser.add_argument("--content-root", default=Path("content"), type=Path)
    args = parser.parse_args()

    violations = validate(args.repo_root, args.content_root)
    if violations:
        print("Docs link validation failed:")
        for violation in violations:
            print(f"- {violation}")
        return 1

    print("Docs link validation passed.")
    print("Anchor validation: deferred; local targets are validated without checking heading anchors.")
    return 0


def validate(repo_root: Path, content_root: Path | str = "content") -> list[str]:
    repo_root = repo_root.resolve()
    content_root_path = Path(content_root)
    if not content_root_path.is_absolute():
        content_root_path = repo_root / content_root_path
    content_root_path = content_root_path.resolve()

    violations: list[str] = []
    registry: doc_registry.DocumentRegistry | None = None
    try:
        registry = doc_registry.build_registry(content_root_path)
    except doc_registry.DocumentRegistryError as exc:
        violations.extend(f"content registry: {violation}" for violation in exc.violations)
        registry = doc_registry.build_registry(content_root_path, validate=False)

    markdown_files = docs_markdown_files(repo_root, content_root_path)
    if registry is not None:
        violations.extend(validate_related_docs(repo_root, markdown_files, registry))
        violations.extend(validate_wiki_refs(repo_root, markdown_files, registry))
        violations.extend(validate_markdown_refs(repo_root, markdown_files, registry))
        violations.extend(validate_task_template_docs(repo_root, registry))
        violations.extend(validate_backend_seed_doc_ids(repo_root, registry))
    else:
        violations.extend(validate_markdown_refs(repo_root, markdown_files, None))

    return violations


def docs_markdown_files(repo_root: Path, content_root: Path) -> list[Path]:
    files: set[Path] = set(path for path in content_root.rglob("*.md") if path.is_file())
    for pattern in PROCESS_MARKDOWN_GLOBS:
        files.update(path for path in repo_root.glob(pattern) if path.is_file())
    for repo_path in PROCESS_MARKDOWN_FILES:
        path = repo_root / repo_path
        if path.exists():
            files.add(path)
    return sorted(path.resolve() for path in files)


def validate_related_docs(
    repo_root: Path,
    markdown_files: list[Path],
    registry: doc_registry.DocumentRegistry,
) -> list[str]:
    violations: list[str] = []
    for markdown_path in markdown_files:
        raw, _body = sop_parse.split_frontmatter(markdown_path.read_text(encoding="utf-8", errors="replace"))
        metadata = sop_parse.parse_frontmatter(raw) if raw else {}
        for related in _strings(metadata.get("related_docs")):
            if _is_external_or_anchor(related):
                continue
            if _registry_resolves(registry, related):
                continue
            if any(candidate.exists() for candidate in _reference_candidates(repo_root, markdown_path, related)):
                continue
            violations.append(f"{_repo_path(repo_root, markdown_path)}: related_docs reference not found: {related!r}")
    return violations


def validate_wiki_refs(
    repo_root: Path,
    markdown_files: list[Path],
    registry: doc_registry.DocumentRegistry,
) -> list[str]:
    violations: list[str] = []
    for markdown_path in markdown_files:
        text = _strip_code_blocks(markdown_path.read_text(encoding="utf-8", errors="replace"))
        for match in WIKI_REF_RE.finditer(text):
            target = match.group("target").strip()
            if not _registry_resolves(registry, target):
                violations.append(f"{_repo_path(repo_root, markdown_path)}: wiki reference not found: {target!r}")
    return violations


def validate_markdown_refs(
    repo_root: Path,
    markdown_files: list[Path],
    registry: doc_registry.DocumentRegistry | None,
) -> list[str]:
    violations: list[str] = []
    for ref in iter_markdown_refs(markdown_files):
        source = _repo_path(repo_root, ref.source_path)
        target = _strip_markdown_title(ref.target)
        if _is_external_or_anchor(target):
            continue
        if target.startswith("doc:"):
            doc_ref, _anchor = _split_link_target(target)
            if registry is not None and not _registry_resolves(registry, doc_ref):
                violations.append(f"{source}: doc reference not found: {target!r}")
            continue
        path_part, _anchor = _split_link_target(target)
        if not path_part:
            continue
        target_path = _resolve_local_link(repo_root, ref.source_path, path_part)
        if not target_path.exists():
            label = "image target" if ref.is_image else "link target"
            violations.append(f"{source}: {label} not found: {target!r}")
    return violations


def iter_markdown_refs(markdown_files: list[Path]) -> list[MarkdownReference]:
    refs: list[MarkdownReference] = []
    for markdown_path in markdown_files:
        text = _strip_code_blocks(markdown_path.read_text(encoding="utf-8", errors="replace"))
        for match in MARKDOWN_LINK_RE.finditer(text):
            refs.append(
                MarkdownReference(
                    source_path=markdown_path,
                    target=match.group("target"),
                    is_image=bool(match.group("image")),
                )
            )
    return refs


def validate_task_template_docs(repo_root: Path, registry: doc_registry.DocumentRegistry) -> list[str]:
    violations: list[str] = []
    templates_dir = repo_root / "content" / "tasks" / "templates"
    for template_path in sorted(templates_dir.glob("*.md")):
        repo_path = _repo_path(repo_root, template_path)
        record = registry.by_path.get(repo_path)
        if record is None:
            violations.append(f"{repo_path}: task template is missing from document registry")
        elif record.doc_type != "task-template":
            violations.append(f"{repo_path}: task template doc_type must be task-template, got {record.doc_type!r}")
    return violations


def validate_backend_seed_doc_ids(repo_root: Path, registry: doc_registry.DocumentRegistry) -> list[str]:
    seed_path = repo_root / "backend" / "scripts" / "seed-templates.ts"
    if not seed_path.exists():
        return ["backend/scripts/seed-templates.ts: seed template source is required"]

    text = seed_path.read_text(encoding="utf-8", errors="replace")
    external_docs = _external_seed_docs(text)
    violations: list[str] = []

    for doc_id, info in external_docs.items():
        if _registry_resolves(registry, doc_id):
            continue
        path = info["path"]
        reason = info["reason"]
        if path.startswith("content/"):
            violations.append(f"{_repo_path(repo_root, seed_path)}: external sourceDocId {doc_id!r} points into content/: {path}")
        if not (repo_root / path).exists():
            violations.append(f"{_repo_path(repo_root, seed_path)}: external sourceDocId {doc_id!r} path not found: {path}")
        if not reason.strip():
            violations.append(f"{_repo_path(repo_root, seed_path)}: external sourceDocId {doc_id!r} needs a reason")

    for doc_id in sorted(_source_doc_ids(text)):
        if doc_id in external_docs:
            continue
        if not _registry_resolves(registry, doc_id):
            violations.append(f"{_repo_path(repo_root, seed_path)}: sourceDocIds reference not found: {doc_id!r}")

    for doc_id in sorted(set(INSTRUCTION_DOC_ID_RE.findall(text))):
        if not _registry_resolves(registry, doc_id):
            violations.append(f"{_repo_path(repo_root, seed_path)}: instructionDocId reference not found: {doc_id!r}")

    return violations


def _source_doc_ids(seed_text: str) -> set[str]:
    doc_ids: set[str] = set()
    for match in SOURCE_DOC_IDS_RE.finditer(seed_text):
        body = match.group("body")
        doc_ids.update(item.group("value") for item in STRING_LITERAL_RE.finditer(body))
    for const_match in re.finditer(r"const\s+(?P<name>[A-Z0-9_]*SOURCE_DOC_IDS)\s*=\s*\[(?P<body>.*?)\];", seed_text, re.DOTALL):
        if "EXTERNAL_SOURCE_DOC_IDS" in const_match.group("name"):
            continue
        doc_ids.update(item.group("value") for item in STRING_LITERAL_RE.finditer(const_match.group("body")))
    return doc_ids


def _external_seed_docs(seed_text: str) -> dict[str, dict[str, str]]:
    return {
        match.group("id"): {"path": match.group("path"), "reason": match.group("reason")}
        for match in EXTERNAL_DOC_OBJECT_RE.finditer(seed_text)
    }


def _strip_code_blocks(markdown: str) -> str:
    return INLINE_CODE_RE.sub("", CODE_BLOCK_RE.sub("", markdown))


def _strip_markdown_title(target: str) -> str:
    target = target.strip()
    if target.startswith("<") and ">" in target:
        return target[1 : target.index(">")].strip()
    if " " in target:
        return target.split()[0].strip()
    return target


def _is_external_or_anchor(target: str) -> bool:
    return (
        not target
        or target.startswith("#")
        or target.startswith("//")
        or target == "..."
        or "..." in target
        or target.startswith("path/to/")
        or (SCHEME_RE.match(target) is not None and not target.startswith("doc:"))
    )


def _split_link_target(target: str) -> tuple[str, str]:
    target = unquote(target.strip())
    path_part = target
    anchor = ""
    if "#" in path_part:
        path_part, anchor = path_part.split("#", 1)
    if "?" in path_part:
        path_part = path_part.split("?", 1)[0]
    return path_part, anchor


def _resolve_local_link(repo_root: Path, markdown_path: Path, target: str) -> Path:
    target = target.replace("\\", "/")
    if target.startswith("/"):
        return (repo_root / target.lstrip("/")).resolve()
    return (markdown_path.parent / target).resolve()


def _reference_candidates(repo_root: Path, markdown_path: Path, target: str) -> list[Path]:
    target = doc_registry.normalize_reference(target)
    path_part, _anchor = _split_link_target(target)
    path_part = path_part.replace("\\", "/").strip()
    if not path_part:
        return []
    return [
        _resolve_local_link(repo_root, markdown_path, path_part),
        (repo_root / path_part.lstrip("/")).resolve(),
    ]


def _registry_resolves(registry: doc_registry.DocumentRegistry, reference: str) -> bool:
    try:
        doc_registry.resolve_reference(registry, reference)
        return True
    except (LookupError, ValueError):
        return False


def _strings(value: object) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(item).strip().strip('"').strip("'") for item in value if str(item).strip()]
    if isinstance(value, str):
        return [value.strip().strip('"').strip("'")] if value.strip() else []
    return [str(value).strip()]


def _repo_path(repo_root: Path, path: Path) -> str:
    return path.resolve().relative_to(repo_root).as_posix()


if __name__ == "__main__":
    raise SystemExit(main())

File: backend/scripts/validate-docs-links.ts
#!/usr/bin/env node

/**
 * Validates internal DataOps document references.
 *
 * Checks that every `related_docs` entry, wiki reference, markdown link, and
 * image target in the documentation corpus resolves, that every task template
 * is registered with the right doc type, and that the doc IDs referenced by
 * `seed-templates.ts` exist.
 *
 * Usage:
 *   npx tsx scripts/validate-docs-links.ts [--repo-root <path>] [--content-root <path>]
 */

import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { isAbsolute, join, relative, resolve, dirname } from 'node:path';

import {
  DocumentRegistryError,
  LookupError,
  buildRegistry,
  normalizeReference,
  resolveReference,
  type DocumentRegistry,
} from '../src/docs/docRegistry';
import { parseFrontmatter, splitFrontmatter } from '../src/docs/sop';

const PROCESS_MARKDOWN_DIRS = ['docs', 'templates'];
const PROCESS_MARKDOWN_FILES = ['.goal-v1.md', 'PROJECT_PLAN.md', 'PORTAL_ANALYSIS.md', 'README.md'];

const MARKDOWN_LINK_RE = /(!)?\[[^\]]*\]\(([^)]+)\)/g;
const WIKI_REF_RE = /\[\[([^\]|]+)(?:\|[^\]]+)?\]\]/g;
const CODE_BLOCK_RE = /```[\s\S]*?```/g;
const INLINE_CODE_RE = /`[^`]+`/g;
const SCHEME_RE = /^[a-z][a-z0-9+.-]*:/i;
const SOURCE_DOC_IDS_RE = /sourceDocIds\s*:\s*\[([\s\S]*?)\]/g;
const INSTRUCTION_DOC_ID_RE = /instructionDocId\s*:\s*['"]([^'"]+)['"]/g;
const STRING_LITERAL_RE = /['"]([^'"]+)['"]/g;
const SOURCE_DOC_IDS_CONST_RE = /const\s+([A-Z0-9_]*SOURCE_DOC_IDS)\s*=\s*\[([\s\S]*?)\];/g;
const EXTERNAL_DOC_OBJECT_RE = /\{\s*id:\s*['"]([^'"]+)['"]\s*,\s*path:\s*['"]([^'"]+)['"]\s*,\s*reason:\s*['"]([^'"]+)['"]\s*,?\s*\}/g;

interface MarkdownReference {
  sourcePath: string;
  target: string;
  isImage: boolean;
}

export function validate(repoRoot: string, contentRoot = 'content'): string[] {
  const root = resolve(repoRoot);
  const content = resolve(isAbsolute(contentRoot) ? contentRoot : join(root, contentRoot));

  const violations: string[] = [];
  let registry: DocumentRegistry;
  try {
    registry = buildRegistry(content);
  } catch (error) {
    if (!(error instanceof DocumentRegistryError)) throw error;
    violations.push(...error.violations.map((violation) => `content registry: ${violation}`));
    registry = buildRegistry(content, false);
  }

  const markdownFiles = docsMarkdownFiles(root, content);
  violations.push(...validateRelatedDocs(root, markdownFiles, registry));
  violations.push(...validateWikiRefs(root, markdownFiles, registry));
  violations.push(...validateMarkdownRefs(root, markdownFiles, registry));
  violations.push(...validateTaskTemplateDocs(root, registry));
  violations.push(...validateBackendSeedDocIds(root, registry));
  return violations;
}

function collectMarkdown(dir: string, out: string[] = []): string[] {
  if (!existsSync(dir)) return out;
  for (const entry of readdirSync(dir)) {
    const path = join(dir, entry);
    if (statSync(path).isDirectory()) collectMarkdown(path, out);
    else if (path.endsWith('.md')) out.push(path);
  }
  return out;
}

function docsMarkdownFiles(repoRoot: string, contentRoot: string): string[] {
  const files = new Set<string>(collectMarkdown(contentRoot).map((path) => resolve(path)));
  for (const dir of PROCESS_MARKDOWN_DIRS) {
    for (const path of collectMarkdown(join(repoRoot, dir))) files.add(resolve(path));
  }
  for (const name of PROCESS_MARKDOWN_FILES) {
    const path = join(repoRoot, name);
    if (existsSync(path)) files.add(resolve(path));
  }
  return [...files].sort();
}

function validateRelatedDocs(repoRoot: string, markdownFiles: string[], registry: DocumentRegistry): string[] {
  const violations: string[] = [];
  for (const path of markdownFiles) {
    const [raw] = splitFrontmatter(readFileSync(path, 'utf8'));
    const metadata = raw ? parseFrontmatter(raw) : {};
    for (const related of asStrings((metadata as Record<string, unknown>).related_docs)) {
      if (isExternalOrAnchor(related)) continue;
      if (registryResolves(registry, related)) continue;
      if (referenceCandidates(repoRoot, path, related).some((candidate) => existsSync(candidate))) continue;
      violations.push(`${repoPath(repoRoot, path)}: related_docs reference not found: '${related}'`);
    }
  }
  return violations;
}

function validateWikiRefs(repoRoot: string, markdownFiles: string[], registry: DocumentRegistry): string[] {
  const violations: string[] = [];
  for (const path of markdownFiles) {
    const text = stripCodeBlocks(readFileSync(path, 'utf8'));
    for (const match of text.matchAll(WIKI_REF_RE)) {
      const target = match[1].trim();
      if (!registryResolves(registry, target)) {
        violations.push(`${repoPath(repoRoot, path)}: wiki reference not found: '${target}'`);
      }
    }
  }
  return violations;
}

function validateMarkdownRefs(repoRoot: string, markdownFiles: string[], registry: DocumentRegistry): string[] {
  const violations: string[] = [];
  for (const ref of iterMarkdownRefs(markdownFiles)) {
    const source = repoPath(repoRoot, ref.sourcePath);
    const target = stripMarkdownTitle(ref.target);
    if (isExternalOrAnchor(target)) continue;
    if (target.startsWith('doc:')) {
      const [docRef] = splitLinkTarget(target);
      if (!registryResolves(registry, docRef)) {
        violations.push(`${source}: doc reference not found: '${target}'`);
      }
      continue;
    }
    const [pathPart] = splitLinkTarget(target);
    if (!pathPart) continue;
    if (!existsSync(resolveLocalLink(repoRoot, ref.sourcePath, pathPart))) {
      violations.push(`${source}: ${ref.isImage ? 'image target' : 'link target'} not found: '${target}'`);
    }
  }
  return violations;
}

function iterMarkdownRefs(markdownFiles: string[]): MarkdownReference[] {
  const refs: MarkdownReference[] = [];
  for (const path of markdownFiles) {
    const text = stripCodeBlocks(readFileSync(path, 'utf8'));
    for (const match of text.matchAll(MARKDOWN_LINK_RE)) {
      refs.push({ sourcePath: path, target: match[2], isImage: Boolean(match[1]) });
    }
  }
  return refs;
}

function validateTaskTemplateDocs(repoRoot: string, registry: DocumentRegistry): string[] {
  const violations: string[] = [];
  const templatesDir = join(repoRoot, 'content', 'tasks', 'templates');
  if (!existsSync(templatesDir)) return violations;
  for (const name of readdirSync(templatesDir).sort()) {
    if (!name.endsWith('.md')) continue;
    const path = repoPath(repoRoot, join(templatesDir, name));
    const record = registry.byPath.get(path);
    if (!record) violations.push(`${path}: task template is missing from document registry`);
    else if (record.doc_type !== 'task-template') {
      violations.push(`${path}: task template doc_type must be task-template, got '${record.doc_type}'`);
    }
  }
  return violations;
}

function validateBackendSeedDocIds(repoRoot: string, registry: DocumentRegistry): string[] {
  const seedPath = join(repoRoot, 'backend', 'scripts', 'seed-templates.ts');
  if (!existsSync(seedPath)) return ['backend/scripts/seed-templates.ts: seed template source is required'];

  const text = readFileSync(seedPath, 'utf8');
  const source = repoPath(repoRoot, seedPath);
  const externalDocs = externalSeedDocs(text);
  const violations: string[] = [];

  for (const [docId, info] of externalDocs) {
    if (registryResolves(registry, docId)) continue;
    if (info.path.startsWith('content/')) {
      violations.push(`${source}: external sourceDocId '${docId}' points into content/: ${info.path}`);
    }
    if (!existsSync(join(repoRoot, info.path))) {
      violations.push(`${source}: external sourceDocId '${docId}' path not found: ${info.path}`);
    }
    if (!info.reason.trim()) {
      violations.push(`${source}: external sourceDocId '${docId}' needs a reason`);
    }
  }

  for (const docId of [...sourceDocIds(text)].sort()) {
    if (externalDocs.has(docId)) continue;
    if (!registryResolves(registry, docId)) {
      violations.push(`${source}: sourceDocIds reference not found: '${docId}'`);
    }
  }

  const instructionIds = new Set([...text.matchAll(INSTRUCTION_DOC_ID_RE)].map((match) => match[1]));
  for (const docId of [...instructionIds].sort()) {
    if (!registryResolves(registry, docId)) {
      violations.push(`${source}: instructionDocId reference not found: '${docId}'`);
    }
  }

  return violations;
}

function sourceDocIds(seedText: string): Set<string> {
  const docIds = new Set<string>();
  for (const match of seedText.matchAll(SOURCE_DOC_IDS_RE)) {
    for (const literal of match[1].matchAll(STRING_LITERAL_RE)) docIds.add(literal[1]);
  }
  for (const match of seedText.matchAll(SOURCE_DOC_IDS_CONST_RE)) {
    if (match[1].includes('EXTERNAL_SOURCE_DOC_IDS')) continue;
    for (const literal of match[2].matchAll(STRING_LITERAL_RE)) docIds.add(literal[1]);
  }
  return docIds;
}

function externalSeedDocs(seedText: string): Map<string, { path: string; reason: string }> {
  const docs = new Map<string, { path: string; reason: string }>();
  for (const match of seedText.matchAll(EXTERNAL_DOC_OBJECT_RE)) {
    docs.set(match[1], { path: match[2], reason: match[3] });
  }
  return docs;
}

function stripCodeBlocks(markdown: string): string {
  return markdown.replace(CODE_BLOCK_RE, '').replace(INLINE_CODE_RE, '');
}

function stripMarkdownTitle(target: string): string {
  const value = target.trim();
  if (value.startsWith('<') && value.includes('>')) return value.slice(1, value.indexOf('>')).trim();
  if (value.includes(' ')) return value.split(/\s+/)[0].trim();
  return value;
}

function isExternalOrAnchor(target: string): boolean {
  return (
    !target
    || target.startsWith('#')
    || target.startsWith('//')
    || target.includes('...')
    || target.startsWith('path/to/')
    || (SCHEME_RE.test(target) && !target.startsWith('doc:'))
  );
}

function splitLinkTarget(target: string): [string, string] {
  let value = decodeURIComponent(target.trim());
  let anchor = '';
  const hash = value.indexOf('#');
  if (hash !== -1) {
    anchor = value.slice(hash + 1);
    value = value.slice(0, hash);
  }
  const query = value.indexOf('?');
  if (query !== -1) value = value.slice(0, query);
  return [value, anchor];
}

function resolveLocalLink(repoRoot: string, markdownPath: string, target: string): string {
  const value = target.replace(/\\/g, '/');
  if (value.startsWith('/')) return resolve(join(repoRoot, value.replace(/^\/+/, '')));
  return resolve(join(dirname(markdownPath), value));
}

function referenceCandidates(repoRoot: string, markdownPath: string, target: string): string[] {
  const [rawPath] = splitLinkTarget(normalizeReference(target));
  const pathPart = rawPath.replace(/\\/g, '/').trim();
  if (!pathPart) return [];
  return [
    resolveLocalLink(repoRoot, markdownPath, pathPart),
    resolve(join(repoRoot, pathPart.replace(/^\/+/, ''))),
  ];
}

function registryResolves(registry: DocumentRegistry, reference: string): boolean {
  try {
    resolveReference(registry, reference);
    return true;
  } catch (error) {
    if (error instanceof LookupError || error instanceof TypeError) return false;
    throw error;
  }
}

function asStrings(value: unknown): string[] {
  if (value === null || value === undefined) return [];
  const items = Array.isArray(value) ? value : [value];
  return items
    .map((item) => String(item).trim().replace(/^["']|["']$/g, ''))
    .filter((item) => item.length > 0);
}

function repoPath(repoRoot: string, path: string): string {
  return relative(repoRoot, resolve(path)).split(/[\\/]/).join('/');
}

function readFlag(name: string, fallback: string): string {
  const index = process.argv.indexOf(name);
  if (index === -1) return fallback;
  const value = process.argv[index + 1];
  if (!value || value.startsWith('--')) throw new Error(`${name} requires a value`);
  return value;
}

if (require.main === module) {
  const repoRoot = readFlag('--repo-root', process.cwd());
  const contentRoot = readFlag('--content-root', 'content');
  const violations = validate(repoRoot, contentRoot);
  if (violations.length > 0) {
    console.error('Docs link validation failed:');
    violations.forEach((violation) => console.error(`- ${violation}`));
    process.exit(1);
  }
  console.log('Docs link validation passed.');
  console.log('Anchor validation: deferred; local targets are validated without checking heading anchors.');
}
